clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2958,47 +2983,171 @@ static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
   return Result;
 }
 
-// Handles reqd_work_group_size, work_group_size_hint and max_work_group_size
+template <typename WorkGroupAttrType>
+void Sema::addIntelSYCLTripleArgFunctionAttr(Decl *D,
+                                             const AttributeCommonInfo &CI,
+                                             Expr *XDim, Expr *YDim,
+                                             Expr *ZDim) {
+  assert(XDim && "Attribute must have an argument.");
+  assert(YDim && "Attribute must have an argument.");
+  assert(ZDim && "Attribute must have an argument.");
+
+  if (!XDim->isInstantiationDependent() || !YDim->isInstantiationDependent() ||
+      !ZDim->isInstantiationDependent()) {
+    Optional<llvm::APSInt> XDimVal =
+        XDim->getIntegerConstantExpr(getASTContext());
+    Optional<llvm::APSInt> YDimVal =
+        YDim->getIntegerConstantExpr(getASTContext());
+    Optional<llvm::APSInt> ZDimVal =
+        ZDim->getIntegerConstantExpr(getASTContext());
+
+    if (!XDimVal) {
+      Diag(XDim->getExprLoc(), diag::err_attribute_argument_type)
+          << CI << AANT_ArgumentIntegerConstant << XDim->getSourceRange();
+      return;
+    }
+
+    if (!YDimVal) {
+      Diag(YDim->getExprLoc(), diag::err_attribute_argument_type)
+          << CI << AANT_ArgumentIntegerConstant << YDim->getSourceRange();
+      return;
+    }
+
+    if (!ZDimVal) {
+      Diag(ZDim->getExprLoc(), diag::err_attribute_argument_type)
+          << CI << AANT_ArgumentIntegerConstant << ZDim->getSourceRange();
+      return;
+    }
+
+    int32_t XDimInt = XDimVal->getSExtValue();
+    if (XDimInt <= 0) {
+      Diag(XDim->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
+      return;
+    }
+
+    int32_t YDimInt = YDimVal->getSExtValue();
+    if (YDimInt <= 0) {
+      Diag(YDim->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
+      return;
+    }
+
+    int32_t ZDimInt = ZDimVal->getSExtValue();
+    if (ZDimInt <= 0) {
+      Diag(ZDim->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
+      return;
+    }
+  }
+
+  D->addAttr(::new (Context) WorkGroupAttrType(Context, CI, XDim, YDim, ZDim));
+}
+
 template <typename WorkGroupAttr>
 static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
   if (D->isInvalidDecl())
     return;
 
   uint32_t WGSize[3];
-  if (AL.getKind() == ParsedAttr::AT_ReqdWorkGroupSize &&
-      AL.getAttributeSpellingListIndex() ==
-          ReqdWorkGroupSizeAttr::CXX11_intel_reqd_work_group_size) {
-    WGSize[1] = ReqdWorkGroupSizeAttr::DefaultYDim;
-    WGSize[2] = ReqdWorkGroupSizeAttr::DefaultZDim;
-  } else if (!checkAttributeNumArgs(S, AL, 3))
+
+  Expr *XDimExpr = AL.getArgAsExpr(0);
+
+  // If no attribute argument is specified, set to default value '1'
+  // in ReqdWorkGroupSizeAttr for only with intel::reqd_work_group_size
+  // spelling.
+  Expr *YDimExpr =
```

Comment:
I'd probably stick this complicated logic into a lambda so you can make the Y and Z dims a bit more easily.

Reply:
Done